### ISSUE TITLE
Permission denied on django collectstatic 

### ISSUE THREAD

**Issue by alexgustafson** [dokku/dokku#6225]
### Description of problem

I use dokku in production all the time. However I usually use a Dockerfile to setup my environment.

I wanted to see if I can save some configuration foo by using the herokuish way to automatically detect my Django setup. 

Using the storage plugin I map the Django static path '/var/www/static' to the hosts "/var/www/{project_name}" path. This works smoothly in Docker based systems.

On deploy, the dokku recognized correctly that this is a Django based project. Dokku runs the django "collectstatic" management command, as it should, with the correct django settings configuration.  The App build process stops with a `Permission denied: '/var/www/static'` PermissionError.

I have checked the folder on the host, that it exists.  Permissions are 777. 

Should this be working? or is the storage plugin only for Docker based setups?

### Steps to reproduce

setup folders

`mkdir -p /var/www/{project}`
`sudo chmod 777 -R /var/www/{project}`

run storage plugin:

`dokku storage:mount {project} /var/www/{project}/static:/var/www/static`

### dokku report $APP_NAME

-----> uname: Linux mishki-testing 5.15.0-79-generic #86-Ubuntu SMP Mon Jul 10 16:07:21 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
-----> memory: 
                      total        used        free      shared  buff/cache   available
       Mem:            3816         275         352          14        3187        3235
       Swap:              0           0           0
-----> docker version: 
       Client: Docker Engine - Community
        Version:           24.0.6
        API version:       1.43
        Go version:        go1.20.7
        Git commit:        ed223bc
        Built:             Mon Sep  4 12:31:44 2023
        OS/Arch:           linux/amd64
        Context:           default
       
       Server: Docker Engine - Community
        Engine:
         Version:          24.0.6
         API version:      1.43 (minimum version 1.12)
         Go version:       go1.20.7
         Git commit:       1a79695
         Built:            Mon Sep  4 12:31:44 2023
         OS/Arch:          linux/amd64
         Experimental:     false
        containerd:
         Version:          1.6.24
         GitCommit:        61f9fd88f79f081d64d6fa3bb1a0dc71ec870523
        runc:
         Version:          1.1.9
         GitCommit:        v1.1.9-0-gccaecfc
        docker-init:
         Version:          0.19.0
         GitCommit:        de40ad0
-----> docker daemon info: 
       Client: Docker Engine - Community
        Version:    24.0.6
        Context:    default
        Debug Mode: true
        Plugins:
         buildx: Docker Buildx (Docker Inc.)
           Version:  v0.11.2
           Path:     /usr/libexec/docker/cli-plugins/docker-buildx
         compose: Docker Compose (Docker Inc.)
           Version:  v2.21.0
           Path:     /usr/libexec/docker/cli-plugins/docker-compose
       
       Server:
        Containers: 1
         Running: 1
         Paused: 0
         Stopped: 0
        Images: 7
        Server Version: 24.0.6
        Storage Driver: overlay2
         Backing Filesystem: extfs
         Supports d_type: true
         Using metacopy: false
         Native Overlay Diff: true
         userxattr: false
        Logging Driver: json-file
        Cgroup Driver: systemd
        Cgroup Version: 2
        Plugins:
         Volume: local
         Network: bridge host ipvlan macvlan null overlay
         Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
        Swarm: inactive
        Runtimes: io.containerd.runc.v2 runc
        Default Runtime: runc
        Init Binary: docker-init
        containerd version: 61f9fd88f79f081d64d6fa3bb1a0dc71ec870523
        runc version: v1.1.9-0-gccaecfc
        init version: de40ad0
        Security Options:
         apparmor
         seccomp
          Profile: builtin
         cgroupns
        Kernel Version: 5.15.0-79-generic
        Operating System: Ubuntu 22.04.3 LTS
        OSType: linux
        Architecture: x86_64
        CPUs: 2
        Total Memory: 3.727GiB
        Name: mishki-testing
        ID: 51541f06-1129-4974-839b-a287fc484efe
        Docker Root Dir: /var/lib/docker
        Debug Mode: false
         File Descriptors: 29
         Goroutines: 39
         System Time: 2023-09-28T11:44:03.69467944Z
         EventsListeners: 0
        Experimental: false
        Insecure Registries:
         127.0.0.0/8
        Live Restore Enabled: false
       
-----> git version: git version 2.34.1
-----> sigil version: 0.10.1build+e443be0
-----> herokuish version: 
       herokuish: v0.7.1
       buildpacks:
         heroku-buildpack-multi     v1.2.0
         heroku-buildpack-ruby      v256
         heroku-buildpack-nodejs    v219
         heroku-buildpack-clojure   v90
         heroku-buildpack-python    v235
         heroku-buildpack-java      v73
         heroku-buildpack-gradle    v39
         heroku-buildpack-scala     v96
         heroku-buildpack-play      v26
         heroku-buildpack-php       v237
         heroku-buildpack-go        v178
         heroku-buildpack-nginx     v24
         buildpack-null             v3
-----> dokku version: dokku version 0.31.4
-----> plugn version: plugn: 0.13.0build+fd5297a
-----> dokku plugins: 
         00_dokku-standard    0.31.4 enabled    dokku core standard plugin
         20_events            0.31.4 enabled    dokku core events logging plugin
         app-json             0.31.4 enabled    dokku core app-json plugin
         apps                 0.31.4 enabled    dokku core apps plugin
         builder              0.31.4 enabled    dokku core builder plugin
         builder-dockerfile   0.31.4 enabled    dokku core builder-dockerfile plugin
         builder-herokuish    0.31.4 enabled    dokku core builder-herokuish plugin
         builder-lambda       0.31.4 enabled    dokku core builder-lambda plugin
         builder-null         0.31.4 enabled    dokku core builder-null plugin
         builder-pack         0.31.4 enabled    dokku core builder-pack plugin
         buildpacks           0.31.4 enabled    dokku core buildpacks plugin
         caddy-vhosts         0.31.4 enabled    dokku core caddy-vhosts plugin
         certs                0.31.4 enabled    dokku core certificate management plugin
         checks               0.31.4 enabled    dokku core checks plugin
         common               0.31.4 enabled    dokku core common plugin
         config               0.31.4 enabled    dokku core config plugin
         cron                 0.31.4 enabled    dokku core cron plugin
         docker-options       0.31.4 enabled    dokku core docker-options plugin
         domains              0.31.4 enabled    dokku core domains plugin
         enter                0.31.4 enabled    dokku core enter plugin
         git                  0.31.4 enabled    dokku core git plugin
         haproxy-vhosts       0.31.4 enabled    dokku core haproxy-vhosts plugin
         logs                 0.31.4 enabled    dokku core logs plugin
         network              0.31.4 enabled    dokku core network plugin
         nginx-vhosts         0.31.4 enabled    dokku core nginx-vhosts plugin
         openresty-vhosts     0.31.4 enabled    dokku core openresty-vhosts plugin
         plugin               0.31.4 enabled    dokku core plugin plugin
         ports                0.31.4 enabled    dokku core ports plugin
         postgres             1.35.0 enabled    dokku postgres service plugin
         proxy                0.31.4 enabled    dokku core proxy plugin
         ps                   0.31.4 enabled    dokku core ps plugin
         registry             0.31.4 enabled    dokku core registry plugin
         repo                 0.31.4 enabled    dokku core repo plugin
         resource             0.31.4 enabled    dokku core resource plugin
         run                  0.31.4 enabled    dokku core run plugin
         scheduler            0.31.4 enabled    dokku core scheduler plugin
         scheduler-docker-local 0.31.4 enabled    dokku core scheduler-docker-local plugin
         scheduler-null       0.31.4 enabled    dokku core scheduler-null plugin
         shell                0.31.4 enabled    dokku core shell plugin
         ssh-keys             0.31.4 enabled    dokku core ssh-keys plugin
         storage              0.31.4 enabled    dokku core storage plugin
         trace                0.31.4 enabled    dokku core trace plugin
         traefik-vhosts       0.31.4 enabled    dokku core traefik-vhosts plugin
=====> lurich app-json information
       App json computed selected:    app.json
       App json global selected:      app.json
       App json selected:             
=====> lurich app information
       App created at:                1695896374
       App deploy source:             
       App deploy source metadata:    
       App dir:                       /home/dokku/lurich
       App locked:                    false
=====> lurich builder information
       Builder build dir:             
       Builder computed build dir:    
       Builder computed selected:     
       Builder global build dir:      
       Builder global selected:       
       Builder selected:              
=====> lurich builder-dockerfile information
       Builder dockerfile computed dockerfile path: Dockerfile               
       Builder dockerfile global dockerfile path: Dockerfile               
       Builder dockerfile dockerfile path:                          
=====> lurich builder-herokuish information
       Builder herokuish computed allowed: true                     
       Builder herokuish global allowed: true                     
       Builder herokuish allowed:                              
=====> lurich builder-lambda information
       Builder lambda computed lambdayml path: lambda.yml               
       Builder lambda global lambdayml path: lambda.yml               
       Builder lambda lambdayml path:                          
=====> lurich builder-pack information
       Builder pack computed projecttoml path: project.toml             
       Builder pack global projecttoml path: project.toml             
       Builder pack projecttoml path:                          
=====> lurich buildpacks information
       Buildpacks computed stack:     gliderlabs/herokuish:latest-22
       Buildpacks global stack:       
       Buildpacks list:               
       Buildpacks stack:              
=====> lurich caddy information
       Caddy image:                   lucaslorentz/caddy-docker-proxy:2.8
       Caddy letsencrypt email:                                
       Caddy letsencrypt server:      https://acme-v02.api.letsencrypt.org/directory
       Caddy log level:               ERROR                    
       Caddy polling interval:        5s                       
       Caddy tls internal:            false                    
=====> lurich ssl information
       Ssl dir:                       /home/dokku/lurich/tls   
       Ssl enabled:                   false                    
       Ssl hostnames:                                          
       Ssl expires at:                                         
       Ssl issuer:                                             
       Ssl starts at:                                          
       Ssl subject:                                            
       Ssl verified:                                           
=====> lurich checks information
       Checks disabled list:          none                     
       Checks skipped list:           none                     
       Checks computed wait to retire: 60                       
       Checks global wait to retire:  60                       
       Checks wait to retire:                                  
=====> lurich docker options information
       Docker options build:          --link dokku.postgres.lurich_db:dokku-postgres-lurich-db 
       Docker options deploy:         --link dokku.postgres.lurich_db:dokku-postgres-lurich-db --restart=on-failure:10 -v /var/www/lurich/media:/var/www/media -v /var/www/lurich/static:/var/www/static 
       Docker options run:            --link dokku.postgres.lurich_db:dokku-postgres-lurich-db -v /var/www/lurich/media:/var/www/media -v /var/www/lurich/static:/var/www/static 
=====> lurich domains information
       Domains app enabled:           true                     
       Domains app vhosts:            lurich.mishki-testing    
       Domains global enabled:        true                     
       Domains global vhosts:         mishki-testing           
=====> lurich git information
       Git deploy branch:             main                     
       Git global deploy branch:      master                   
       Git keep git dir:              false                    
       Git rev env var:               GIT_REV                  
       Git sha:                                                
       Git source image:                                       
       Git last updated at:                                    
=====> lurich haproxy information
       Haproxy image:                 byjg/easy-haproxy:4.3.0  
       Haproxy letsencrypt email:                              
       Haproxy letsencrypt server:    https://acme-v02.api.letsencrypt.org/directory
       Haproxy log level:             ERROR                    
=====> lurich logs information
       Logs computed max size:        10m
       Logs global max size:          10m
       Logs global vector sink:       
       Logs max size:                 
       Logs vector sink:              
=====> lurich network information
       Network attach post create:           
       Network attach post deploy:           
       Network bind all interfaces:          false
       Network computed attach post create:  
       Network computed attach post deploy:  
       Network computed bind all interfaces: false
       Network computed initial network:     
       Network computed tld:                 
       Network global attach post create:    
       Network global attach post deploy:    
       Network global bind all interfaces:   false
       Network global initial network:       
       Network global tld:                   
       Network initial network:              
       Network static web listener:          
       Network tld:                          
       Network web listeners:                
=====> lurich nginx information
       Nginx access log format:                                
       Nginx access log path:         /var/log/nginx/lurich-access.log
       Nginx bind address ipv4:                                
       Nginx bind address ipv6:       ::                       
       Nginx client max body size:                             
       Nginx disable custom config:   false                    
       Nginx error log path:          /var/log/nginx/lurich-error.log
       Nginx global hsts:             true                     
       Nginx computed hsts:           true                     
       Nginx hsts:                                             
       Nginx hsts include subdomains: true                     
       Nginx hsts max age:            15724800                 
       Nginx hsts preload:            false                    
       Nginx computed nginx conf sigil path: nginx.conf.sigil         
       Nginx global nginx conf sigil path: nginx.conf.sigil         
       Nginx nginx conf sigil path:                            
       Nginx proxy buffer size:       4096                     
       Nginx proxy buffering:         on                       
       Nginx proxy buffers:           8 4096                   
       Nginx proxy busy buffers size: 8192                     
       Nginx proxy read timeout:      60s                      
       Nginx last visited at:                                  
       Nginx x forwarded for value:   $remote_addr             
       Nginx x forwarded port value:  $server_port             
       Nginx x forwarded proto value: $scheme                  
       Nginx x forwarded ssl:                                  
=====> lurich openresty information
       Openresty access log format:                            
       Openresty access log path:     /var/log/nginx/lurich-access.log
       Openresty bind address ipv4:                            
       Openresty bind address ipv6:   ::                       
       Openresty client max body size:                          
       Openresty error log path:      /var/log/nginx/lurich-error.log
       Openresty global hsts:         true                     
       Openresty computed hsts:       true                     
       Openresty hsts:                                         
       Openresty hsts include subdomains: true                     
       Openresty hsts max age:        15724800                 
       Openresty hsts preload:        false                    
       Openresty image:               dokku/openresty-docker-proxy:0.5.6
       Openresty letsencrypt email:                            
       Openresty letsencrypt server:  https://acme-v02.api.letsencrypt.org/directory
       Openresty proxy buffer size:   4096                     
       Openresty proxy buffering:     on                       
       Openresty proxy buffers:       8 4096                   
       Openresty proxy busy buffers size: 8192                     
       Openresty proxy read timeout:  60s                      
       Openresty x forwarded for value: $remote_addr             
       Openresty x forwarded port value: $server_port             
       Openresty x forwarded proto value: $scheme                  
       Openresty x forwarded ssl:                              
=====> lurich ports information
       Ports map:                     
       Ports map detected:            http:80:5000
=====> lurich proxy information
       Proxy enabled:                 true
       Proxy type:                    nginx
=====> lurich ps information
       Deployed:                      false
       Processes:                     0
       Ps can scale:                  true
       Ps computed procfile path:     Procfile
       Ps global procfile path:       Procfile
       Ps procfile path:              
       Ps restart policy:             on-failure:10
       Restore:                       true
       Running:                       false
=====> lurich registry information
       Registry computed image repo:      dokku/lurich
       Registry computed push on release: false
       Registry computed server:          
       Registry global push on release:   
       Registry global server:            
       Registry image repo:               
       Registry push on release:          
       Registry server:                   
       Registry tag version:              
=====> lurich resource information
=====> lurich scheduler information
       Scheduler computed selected:   docker-local
       Scheduler global selected:     docker-local
       Scheduler selected:            
=====> lurich scheduler-docker-local information
       Scheduler docker local disable chown:                          
       Scheduler docker local init process: true                     
       Scheduler docker local parallel schedule count:                          
=====> lurich storage information
       Storage build mounts:                                   
       Storage deploy mounts:         -v /var/www/lurich/media:/var/www/media -v /var/www/lurich/static:/var/www/static 
       Storage run mounts:            -v /var/www/lurich/media:/var/www/media -v /var/www/lurich/static:/var/www/static 
=====> lurich traefik information
       Traefik api enabled:           false                    
       Traefik api vhost:             traefik.dokku.me         
       Traefik basic auth password:                            
       Traefik basic auth username:                            
       Traefik dashboard enabled:     false                    
       Traefik image:                 traefik:v2.10            
       Traefik letsencrypt email:                              
       Traefik letsencrypt server:    https://acme-v02.api.letsencrypt.org/directory
       Traefik log level:             ERROR                    
       Traefik priority:

**Comment by josegonzalez** [comment]
Are container processes running as root, or as a different user? That other user maps to some UID/GID on the host, and that needs to be set as the owner of the host directory. You can check the UID/GID by running `ps auxf` on the host and seeing what your app processes map to.

**Comment by alexgustafson** [comment]
dockerd and containerd are running as root. Actually everything on the host is running as root, with the exception of some unrelated processes ( systemd, sshd, dbus, etc )

If you could confirm that the storage plugin should work with the herokuish build setup, and the paths look ok then my next step would be to wipe the server and start again from scratch. 

Maybe the internal part of the path should map to something within the workspace of the container rather than /var/www ? This has never been an issue with Docker based deployments, so I am a bit unsure.

**Comment by josegonzalez** [comment]
Herokuish processes run as `32767:32767` - as documented [here](https://dokku.com/docs/advanced-usage/persistent-storage/#creating-storage-directories) - so the folder on the host has to be chowned by that uid/gid in order to be written from within the container. Many dockerfile-based deploys run as root, so thats not typically an issue (since root is the same on the host and the container).

This functionality otherwise works fine for herokuish, you just need to set permissions correctly on the host folder.

**Comment by alexgustafson** [comment]
Ok, so setting chown to 32767:32787 on /var/www/{project} recursively didnt solve my problem.

What did work was sticking to the documentation more closely. I mapped the container's internal /app/storage directory to the host's `var/lib/dokku/data/storage/{project}` path. 

Then I ran the `ensure-directory` command.

After that everything my deployment worked. ( now I have to see if the new path works for nginx as well, but I will leave that for another ticket )

Thanks for the hand-holding!
